src/codegen.cpp

```diff
@@ -4146,6 +4146,22 @@ static LLVMValueRef ir_render_frame_address(CodeGen *g, IrExecutable *executable
     return LLVMBuildCall(g->builder, get_frame_address_fn_val(g), &zero, 1, "");
 }
 
+static LLVMValueRef ir_render_handle(CodeGen *g, IrExecutable *executable,
+        IrInstructionHandle *instruction)
+{
+
+    bool is_async = executable->fn_entry != nullptr && 
+        executable->fn_entry->type_entry->data.fn.fn_type_id.cc == CallingConventionAsync;
+
+    if (!is_async || !executable->coro_handle) {
+      add_node_error(g, instruction->base.source_node, buf_sprintf("@handle() in non-async function"));
```

Comment:
in render functions in codegen.cpp it is too late - we need to put this semantic analysis and compile error in ir.cpp in the analyze_instruction_handle function. That's why that assert was there that you commented out :-)